src/dataflow/src/source/mod.rs

```diff
@@ -849,6 +853,221 @@ enum MessageProcessing {
+        if *timestamp < now {
+            *timestamp = now;
+            self.sender
+                .send(Event::Progress(Some(*timestamp)))
+                .await
+                .or_else(|_| Err(anyhow!("channel closed")))?;
+        }
+        Ok(())
+    }
+}
+
+/// Simple sources must implement this trait. Sources will then get created as part of the
+/// [`create_source_simple`] function.
+///
+/// Each simple source is given access to a timestamper instance that can be used to insert or
+/// retract rows for this source. See the API of the [Timestamper](Timestamper) for more details.
+#[async_trait]
+pub trait SimpleSource {
+    /// Consumes the instance of this SimpleSource and converts it into an async state machine that
+    /// submits rows using the provided timestamper.
+    ///
+    /// Implementors should return an Err(_) if an unrecoverable error is encountered or Ok(()) when
+    /// they have finished consuming the upstream data.
+    async fn start(self, timestamper: &Timestamper) -> Result<(), SourceError>;
+}
+
+/// Creates a source dataflow operator from a connector implementing [SimpleSource](SimpleSource)
+pub fn create_source_simple<G, C>(
+    config: SourceConfig<G>,
+    connector: C,
+) -> (
+    (
+        timely::dataflow::Stream<G, (Row, Timestamp, Diff)>,
+        timely::dataflow::Stream<G, SourceError>,
+    ),
+    Option<SourceToken>,
+)
+where
+    G: Scope<Timestamp = Timestamp>,
+    C: SimpleSource + Send + 'static,
+{
+    let SourceConfig {
+        name,
+        scope,
+        active,
+        timestamp_frequency,
+        ..
+    } = config;
+
+    let (tx, mut rx) = mpsc::channel(64);
+
+    if active {
+        tokio::spawn(async move {
+            let timestamper = Timestamper::new(tx, timestamp_frequency);
+            let source = connector.start(&timestamper);
+            tokio::pin!(source);
+
+            loop {
+                tokio::select! {
+                    res = timestamper.tick() => {
+                        if res.is_err() {
+                            break;
+                        }
+                    }
+                    res = &mut source => {
+                        if let Err(err) = res {
+                            let _ = timestamper.error(err).await;
+                        }
+                        break;
+                    }
+                }
+            }
+        });
+    }
+
+    let (stream, capability) = source(scope, name, move |info| {
+        let activator = Arc::new(scope.sync_activator_for(&info.address[..]));
+
+        move |cap, output| {
+            if !active {
+                return SourceStatus::Done;
+            }
+
+            let waker = futures::task::waker_ref(&activator);
+            let mut context = Context::from_waker(&waker);
+
+            while let Poll::Ready(item) = rx.poll_recv(&mut context) {
+                match item {
+                    Some(Event::Progress(None)) => {}
```

Comment:
hmmm it seems like we never actually send a None progress so maybe we ought to mark this unreachable? (idk I don't feel too strongly about it because it is riskier error handling). Also, did you use `Option<Timestamp>`because of the `IntoIterator` trait bound or something else?

Reply:
yeah, I used `Option<Timestamp>` because the `ToStreamAsync` trait in timely needs it. I agree with your suggestion though, I'll mark that path as unreachable